notebook/gateway/managers.py

```diff
@@ -17,37 +17,47 @@
 
 from traitlets import Instance, Unicode, Float, Bool, default, validate, TraitError
 from traitlets.config import SingletonConfigurable
+from ..db_util import ExecuteQueries
 
 
 class GatewayClient(SingletonConfigurable):
-    """This class manages the configuration.  It's its own singleton class so that we
-       can share these values across all objects.  It also contains some helper methods
-        to build request arguments out of the various config options.
-
+    """
+    This class manages the configuration. It's its own singleton class so that we can share these values across
+    all objects. It also contains some helper methods to build request arguments out of the various config options.
     """
 
-    url = Unicode(default_value=None, allow_none=True, config=True,
+    urls = Unicode(default_value=None, allow_none=True, config=True,
         help="""The url of the Kernel or Enterprise Gateway server where
         kernel specifications are defined and kernel management takes place.
         If defined, this Notebook server acts as a proxy for all kernel
         management and kernel specification retrieval.  (JUPYTER_GATEWAY_URL env var)
         """
     )
 
-    url_env = 'JUPYTER_GATEWAY_URL'
+    ml_nodes = ExecuteQueries().get_mlnodes()
+    ml_node_urls = []
+    for record in ml_nodes:
+        ml_node_urls.append(str(record.ip_address) + ':8888')
+
+    urls = ml_node_urls
+
+    urls_env = 'JUPYTER_GATEWAY_URL'
 
-    @default('url')
+    @default('urls')
     def _url_default(self):
         return os.environ.get(self.url_env)
 
-    @validate('url')
+    @validate('urls')
     def _url_validate(self, proposal):
-        value = proposal['value']
-        # Ensure value, if present, starts with 'http'
-        if value is not None and len(value) > 0:
-            if not str(value).lower().startswith('http'):
-                raise TraitError("GatewayClient url must start with 'http': '%r'" % value)
-        return value
+        values = []
+        for url in proposal:
+            value = url['value']
```

Comment:
Is ```value``` always going to be present ?

Reply:
This is not being used anywhere. Updated the flow. But have just kept the function for later references.